examples/vmod_event/src/lib.rs

```diff
@@ -1,38 +1,62 @@
-varnish::boilerplate!();
+#[varnish::vmod]
+mod event {
+    use std::sync::atomic::AtomicI64;
+    use std::sync::atomic::Ordering::Relaxed;
 
-use varnish::vcl::ctx::{Ctx, Event, LogTag};
-use varnish::vcl::vpriv::VPriv;
+    use varnish::vcl::ctx::{Ctx, Event, LogTag};
 
-varnish::vtc!(test01);
+    /// number of times on_event() was called with `Event::Load`
+    /// `on_event()` will increment it and store it in the `PRIV_VCL`
+    /// loaded() will access it and return to the VCL
+    static EVENT_LOADED_COUNT: AtomicI64 = AtomicI64::new(0);
 
-// number of times event() was run with Event::Load
-// event will increment it and store a VCL-local copy via VPriv, that loaded() will be able to
-// retrieve
-static mut N: i64 = 0;
+    pub fn loaded(#[shared_per_vcl] shared: Option<&i64>) -> i64 {
+        shared.copied().unwrap_or(0)
+    }
 
-pub fn loaded(_: &Ctx, vp: &VPriv<i64>) -> i64 {
-    // unwrapping is safe here as we know that the event function will have loaded something
-    *vp.as_ref().unwrap()
-}
+    /// FIXME: does Varnish guarantee that `loaded()` will never be called while `on_event()` is running?
+    /// because we are touching a `static mut` variable, this should be highly unsafe, however, Varnish
+    /// guarantees that event functions are called sequentially in the cli thread, so we'll be fine
+    #[event]
+    pub fn on_event(
+        ctx: &mut Ctx,
+        #[shared_per_vcl] shared: &mut Option<Box<i64>>,
+        event: Event,
+    ) -> Result<i64, &'static str> {
+        // log the event, showing that it implements Debug
+        ctx.log(LogTag::Debug, format!("event: {event:?}"));
+
+        // we only care about load events, which is why we don't use `match`
+        if matches!(event, Event::Load) {
+            // increment the count in a thread-safe way
+            let last_count = EVENT_LOADED_COUNT.fetch_add(1, Relaxed);
+            if last_count == 1 {
+                // FIXME: why, what reasons?
+                // fail the second load, because reasons
+                // FIXME: what should VCL wrapper return in case of an error?
```

Comment:
don't you like my string?

Reply:
lol, i updated it a bit now that i understand its for demoing (for a moment i thought it was some internal idiosyncrasy